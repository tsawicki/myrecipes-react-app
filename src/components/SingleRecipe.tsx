import React from "react";
import { Recipe } from "../models/Recipe.model";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faEdit, faTrash } from "@fortawesome/free-solid-svg-icons";

interface OwnProps {
  recipe: Recipe;
  isExpanded: boolean;
  setExpanded: (recipeId: string) => void;
  onEdit: (recipeId: string) => void;
  onDelete: (recipeId: string) => void;
}

const SingleRecipe: React.FunctionComponent<OwnProps> = ({
  recipe,
  isExpanded,
  setExpanded,
  onEdit,
  onDelete
}: OwnProps) => {
  // const [isExpanded, setIsExpanded] = useState(false);

  const EditLink = () => (
    <a
      className="small-link"
      href="#edit"
      onClick={event => {
        event.preventDefault();
        onEdit(recipe.id);
      }}
      data-testid="edit-link"
    >
      <FontAwesomeIcon
        icon={faEdit}
        size="lg"
        style={{ marginRight: "10px" }}
      />
      edit
    </a>
  );

  const DeleteLink = () => (
    <a
      className="small-link"
      href="#delete"
      onClick={event => {
        event.preventDefault();
        onDelete(recipe.id);
      }}
      data-testid="delete-link"
    >
      <FontAwesomeIcon
        icon={faTrash}
        size="lg"
        style={{ marginRight: "10px" }}
      />
      delete
    </a>
  );

  return (
    <div className="recipe-container">
      <h2
        className="recipe-title"
        data-testid="recipe-title"
        onClick={() => setExpanded(recipe.id)}
      >
        {recipe.name}
      </h2>
      {isExpanded && (
        <>
          <div className="links-container">
            <EditLink />
            <DeleteLink />
          </div>
          <h5 className="section-header">Ingredients</h5>
          <ul>
            {recipe.ingredients.split(", ").map((ingredient, index) => (
              <li key={index}>{ingredient}</li>
            ))}
          </ul>
          {recipe.directions && (
            <div className="directions-section">
              <h5 className="section-header">Directions</h5>
              <div>{recipe.directions}</div>
            </div>
          )}
        </>
      )}
    </div>
  );
};

export default SingleRecipe;
