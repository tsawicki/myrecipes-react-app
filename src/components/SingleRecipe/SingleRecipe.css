.recipe-container {
  border: 1px solid #dedede;
  padding: 20px;
  transition: all 0.4s;
  background: #fff;
  border-radius: 5px;
  margin-bottom: 1px;
}

.recipe-container:hover {
  box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, 0.05);
  position: relative;
}

.directions-section {
  white-space: pre-wrap;
}

h5.section-header {
  margin-bottom: 8px;
}

.links-container {
  padding-top: 5px;
}

h2.recipe-title {
  margin: 0;
  cursor: pointer;
}
