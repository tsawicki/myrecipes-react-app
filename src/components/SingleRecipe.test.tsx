import React from "react";
import { render } from "@testing-library/react";
import SingleRecipe from "./SingleRecipe";

test("SingleRecipe component should match snapshot", () => {
  describe("for not extended component", () => {
    const wrapper = render(
      <SingleRecipe
        recipe={{
          id: "1234",
          name: "test recipe",
          ingredients: "test ingredients",
          directions: ""
        }}
        isExpanded={false}
        setExpanded={jest.fn()}
        onEdit={jest.fn()}
        onDelete={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  describe("for not extended component", () => {
    const wrapper = render(
      <SingleRecipe
        recipe={{
          id: "1234",
          name: "test recipe",
          ingredients: "test ingredients",
          directions: "test directions"
        }}
        isExpanded={true}
        setExpanded={() => {}}
        onEdit={() => {}}
        onDelete={() => {}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});

test("SingleRecipe should call onEdit function on link click", () => {
  const onEdit = jest.fn();
  const wrapper = render(
    <SingleRecipe
      recipe={{
        id: "1234",
        name: "test recipe",
        ingredients: "test ingredients",
        directions: ""
      }}
      isExpanded={true}
      setExpanded={() => {}}
      onEdit={onEdit}
      onDelete={() => {}}
    />
  );
  wrapper.getByTestId("edit-link").click();
  expect(onEdit).toHaveBeenCalledWith("1234");
});

test("SingleRecipe should call onEdit function on link click", () => {
  const onDelete = jest.fn();
  const wrapper = render(
    <SingleRecipe
      recipe={{
        id: "1234",
        name: "test recipe",
        ingredients: "test ingredients",
        directions: ""
      }}
      isExpanded={true}
      setExpanded={() => {}}
      onEdit={() => {}}
      onDelete={onDelete}
    />
  );
  wrapper.getByTestId("delete-link").click();
  expect(onDelete).toHaveBeenCalledWith("1234");
});

test("SingleRecipe should call setExpanded on header click", () => {
  const setExpanded = jest.fn();
  const wrapper = render(
    <SingleRecipe
      recipe={{
        id: "1234",
        name: "test recipe",
        ingredients: "test ingredients",
        directions: ""
      }}
      isExpanded={true}
      setExpanded={setExpanded}
      onEdit={() => {}}
      onDelete={() => {}}
    />
  );
  wrapper.getByTestId("recipe-title").click();
  expect(setExpanded).toHaveBeenCalledWith("1234");
});
