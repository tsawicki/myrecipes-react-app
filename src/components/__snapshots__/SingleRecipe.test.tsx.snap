// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleRecipe component should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="recipe-container"
      >
        <h2
          class="recipe-title"
          data-testid="recipe-title"
        >
          test recipe
        </h2>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="recipe-container"
    >
      <h2
        class="recipe-title"
        data-testid="recipe-title"
      >
        test recipe
      </h2>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SingleRecipe component should match snapshot 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="recipe-container"
      >
        <h2
          class="recipe-title"
          data-testid="recipe-title"
        >
          test recipe
        </h2>
      </div>
    </div>
    <div>
      <div
        class="recipe-container"
      >
        <h2
          class="recipe-title"
          data-testid="recipe-title"
        >
          test recipe
        </h2>
        <div
          class="links-container"
        >
          <a
            class="small-link"
            data-testid="edit-link"
            href="#edit"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-edit fa-w-18 fa-lg "
              data-icon="edit"
              data-prefix="fas"
              focusable="false"
              role="img"
              style="margin-right: 10px;"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
                fill="currentColor"
              />
            </svg>
            edit
          </a>
          <a
            class="small-link"
            data-testid="delete-link"
            href="#delete"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-trash fa-w-14 fa-lg "
              data-icon="trash"
              data-prefix="fas"
              focusable="false"
              role="img"
              style="margin-right: 10px;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                fill="currentColor"
              />
            </svg>
            delete
          </a>
        </div>
        <h5
          class="section-header"
        >
          Ingredients
        </h5>
        <ul>
          <li>
            test ingredients
          </li>
        </ul>
        <div
          class="directions-section"
        >
          <h5
            class="section-header"
          >
            Directions
          </h5>
          <div>
            test directions
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="recipe-container"
    >
      <h2
        class="recipe-title"
        data-testid="recipe-title"
      >
        test recipe
      </h2>
      <div
        class="links-container"
      >
        <a
          class="small-link"
          data-testid="edit-link"
          href="#edit"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-edit fa-w-18 fa-lg "
            data-icon="edit"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="margin-right: 10px;"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
              fill="currentColor"
            />
          </svg>
          edit
        </a>
        <a
          class="small-link"
          data-testid="delete-link"
          href="#delete"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-trash fa-w-14 fa-lg "
            data-icon="trash"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="margin-right: 10px;"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              fill="currentColor"
            />
          </svg>
          delete
        </a>
      </div>
      <h5
        class="section-header"
      >
        Ingredients
      </h5>
      <ul>
        <li>
          test ingredients
        </li>
      </ul>
      <div
        class="directions-section"
      >
        <h5
          class="section-header"
        >
          Directions
        </h5>
        <div>
          test directions
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
